ironfish-cli/src/commands/service/snapshot.ts

```diff
@@ -0,0 +1,213 @@
+    if (!bucket) {
+      this.log(
+        `Cannot upload snapshot without bucket URL. You must set IRONFISH_SNAPSHOT_BUCKET or pass --bucket flag.`,
+      )
+      this.exit(1)
+    }
+
+    let exportDir
+
+    if (flags.path) {
+      exportDir = this.sdk.fileSystem.resolve(flags.path)
+    } else {
+      try {
+        exportDir = await fsAsync.mkdtemp(`${os.tmpdir()}${path.sep}`)
+      } catch (err) {
+        this.log(`Could not create temp folder for snapshot generation`)
+        this.exit(1)
+      }
+    }
+    Assert.isNotUndefined(exportDir)
+
+    const blockExportPath = this.sdk.fileSystem.join(exportDir, 'blocks')
+    await this.sdk.fileSystem.mkdir(blockExportPath, { recursive: true })
+
+    this.log('Connecting to node...')
+
+    const client = await this.sdk.connectRpc()
+
+    const response = client.snapshotChainStream({
+      start: args.start as number | null,
+      stop: args.stop as number | null,
+      maxBlocksPerChunk: flags.maxBlocksPerChunk,
+    })
+
+    const { start, stop } = await AsyncUtils.first(response.contentStream())
+    this.log(`Retrieving blocks from ${start} -> ${stop} for snapshot generation`)
+
+    const progress = CliUx.ux.progress({
+      format: 'Retrieving blocks: [{bar}] {value}/{total} {percentage}% | ETA: {eta}s',
+    }) as ProgressBar
+
+    progress.start(stop - start + 1, 0)
+
+    for await (const result of response.contentStream()) {
+      if (result.buffer && result.seq) {
+        const blockFilePath = this.sdk.fileSystem.join(blockExportPath, `${result.seq}`)
+        await fsAsync.writeFile(blockFilePath, Buffer.from(result.buffer))
+        progress.update(result.seq || 0)
+      }
+    }
+
+    progress.stop()
+
+    const timestamp = Date.now()
+
+    const snapshotFileName = `ironfish_snapshot_${timestamp}.tar.gz`
+    const snapshotPath = this.sdk.fileSystem.join(exportDir, snapshotFileName)
+
+    this.log(`Zipping\n    SRC ${blockExportPath}\n    DST ${snapshotPath}\n`)
+    CliUx.ux.action.start(`Zipping ${blockExportPath}`)
+    await this.zipDir(blockExportPath, snapshotPath)
+    const stat = await fsAsync.stat(snapshotPath)
+    const fileSize = stat.size
+    CliUx.ux.action.stop(`done (${FileUtils.formatFileSize(fileSize)})`)
+
+    const hasher = crypto.createHash('sha256')
+    const fileHandle = await fsAsync.open(snapshotPath, 'r')
+    const stream = fileHandle.createReadStream()
+
+    CliUx.ux.action.start(`Creating checksum for ${snapshotPath}`)
+    for await (const data of stream) {
+      hasher.update(data)
+    }
+    const checksum = hasher.digest().toString('hex')
+    CliUx.ux.action.stop(`done`)
+
+    const blockHeight = stop
+
+    CliUx.ux.action.start(`Uploading to ${bucket}`)
+    await this.uploadToBucket(snapshotPath, bucket, 'application/x-compressed-tar')
+    CliUx.ux.action.stop(`done`)
+
+    await fsAsync.writeFile(
+      path.join(exportDir, 'manifest.json'),
```

Comment:
Is it always best to have one huge snapshot? This would suggest so. Seems possible that the manifest should have flexibility to point to multiple files that together make up the current snapshot.

Reply:
I wouldn't say it's better one way or the other; it's just the implementation that I spec'd out. I'm open to creating separate archives so that a user's node will only download the subset of the chain they are missing.

Reply:
But generally - we allow start/stop parameters in the actual command, but then have a manifest setup that ignores those params. 

For instance, I'm seeing that any user could corrupt all imports by simply running the export command with a start param other than 1. This would run, upload to the bucket, then set manifest to point to that new (incomplete) export.

Reply:
Sure, I see your point. I think it's safe to assume that a snapshot should be a complete representation of the main chain, and if we choose to do multiple snapshot files that are described in the manifest, then I think it's also safe to do away with the start and stop parameters as the import command should handle retrieval of the files that contain the latest un-added block. What do you think?

Reply:
I think I agree, though will have to see the implementation as there is a bit of room for ambiguity here.